lib/taskRunner.js

```diff
@@ -34,97 +34,165 @@ util.inherits(TaskRunner, EventEmitter);
+  var runnerFn = this.runInFork ? this._runForked : this._runNormal
+
+  var runSpecHelper = function() {
+
+    if (retries > 0) {
+      taskLogger.log('Task failed, retrying (' + retries + '/' + maxRetries + ')');
+      taskLogger.flush()
+    }
+
+    retries++
+
+    // TODO: reinit runResults here?
+    var runResults = {
+      taskId: self.task.taskId,
+      specs: self.task.specs,
+      capabilities: self.task.capabilities,
+      // The following are populated while running the test:
+      failedCount: 0,
+      exitCode: -1,
+      specResults: []
+    };
+
+    runnerFn.call(self, runResults).then(function() {
+      // console.log('runnerFn then, retries: %d, failed: %d, exit code: %d', retries, runResults.failedCount, runResults.exitCode);
+
+      if ((runResults.failedCount > 0 || runResults.exitCode == 1) && retries <= maxRetries)
+        return runSpecHelper()
+
+      // TaskLogger.flush prints task output to the screen. The reason we don't do it in `_runForked` is
+      // we want to avoid non-sense output. In case if task failed, retried and succedeed we don't want to
+      // pring any information about errors.
+      runResults.taskLogger.flush()
+      deferred.resolve(runResults)
+
+    }).catch(function(err) {
+      taskLogger.log('Protractor runtime error, very likely it is not critical:' + err);
+      taskLogger.flush()
+
+      if (retries <= maxRetries)
+        return runSpecHelper()
+
+      runResults.taskLogger.flush()
+      deferred.reject(err)
     })
-    .on('exit', function(code) {
-      taskLogger.flush();
-      runResults.exitCode = code;
-      deferred.resolve(runResults);
-    });
-
-    childProcess.send({
-      command: 'run',
-      configFile: this.configFile,
-      additionalConfig: this.additionalConfig,
-      capabilities: this.task.capabilities,
-      specs: this.task.specs
-    });
-
-    return deferred.promise;
-  } else {
-    var ConfigParser = require('./configParser');
-    var configParser = new ConfigParser();
-    if (this.configFile) {
-      configParser.addFileConfig(this.configFile);
+  }
+
+  runSpecHelper();
+
+  return deferred.promise
+};
+
+
+TaskRunner.prototype._runForked = function(runResults) {
+  var deferred = q.defer();
+  var self = this;
+
+  var childProcess = child.fork(
+    __dirname + '/runnerCli.js',
+    process.argv.slice(2), {
+      cwd: process.cwd(),
+      silent: true
     }
-    if (this.additionalConfig) {
-      configParser.addConfig(this.additionalConfig);
+  );
+  var taskLogger = new TaskLogger(this.task, childProcess.pid);
+  // I know it's ugly
+  runResults.taskLogger = taskLogger;
+
+  // stdout pipe
+  childProcess.stdout.on('data', function(data) {
+    taskLogger.log(data);
+  });
+
+  // stderr pipe
+  childProcess.stderr.on('data', function(data) {
+    taskLogger.log(data);
+  });
+
+  childProcess.on('message', function(m) {
+    // console.log('zzz msg', self.task.taskId, m.event);
```

Comment:
I want to keep my debugging console.logs for time being. I'll send a change request to original protractor repo once my changes teste